src/driver/test_point_aggregator.py

```diff
@@ -0,0 +1,130 @@
+# as output a scalar value (could be float, int or string).
+AGGREGATORS = {
+    # Number of test points
+    "npoints": lambda x, y: len(x),
+    # Mean error (bias)
+    "mean_error": lambda x, y: np.mean(_reach(x) - _reach(y)),
+    # Mean absolute error in predicted reach
+    "mean_abs_error": lambda x, y: np.mean(np.abs(_reach(x) - _reach(y))),
+    # Mean squared error in predicted reach
+    "mean_squared_error": lambda x, y: np.mean((_reach(x) - _reach(y)) ** 2),
+    # Mean absolute relative error in predicted reach
+    "mean_abs_relative_error": lambda x, y: np.mean(
+        np.abs(_reach(x) - _reach(y)) / _reach(x)
+    ),
+    # Mean squared relative error in predicted reach
+    "mean_squared_relative_error": lambda x, y: np.mean(
+        (_reach(x) - _reach(y)) ** 2 / _reach(x)
+    ),
+    # Error variance in predicted reach
+    "var_error": lambda x, y: np.var(_reach(x) - _reach(y)),
+    # Relative error variance in predicted reach
+    "var_relative_error": lambda x, y: np.var((_reach(x) - _reach(y)) / _reach(x)),
+    # Quantiles of relative error in predicted reach
+    "relative_error_q10": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.10
+    ),
+    "relative_error_q20": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.20
+    ),
+    "relative_error_q30": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.30
+    ),
+    "relative_error_q40": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.40
+    ),
+    "relative_error_q50": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.50
+    ),
+    "relative_error_q60": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.60
+    ),
+    "relative_error_q70": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.70
+    ),
+    "relative_error_q80": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.80
+    ),
+    "relative_error_q90": lambda x, y: np.quantile(
+        np.abs(_reach(x) - _reach(y)) / _reach(x), 0.90
+    ),
+    # Mean shuffle distance
+    "mean_shuffle_distance": lambda x, y: np.mean(
+        [_shuffle_distance(x[i], y[i]) for i in range(len(x))]
+    ),
+    # Mean squared shuffle distance
+    "mean_squared_shuffle_distance": lambda x, y: np.mean(
+        [_shuffle_distance(x[i], y[i]) ** 2 for i in range(len(x))]
+    ),
+    # Variance of shuffle distance
+    "var_shuffle_distance": lambda x, y: np.var(
+        [_shuffle_distance(x[i], y[i]) for i in range(len(x))]
+    ),
+}
+
+
+def _reach(point_list: List[ReachPoint], k=1) -> np.array:
+    """Returns list of k+ frequencies from list of ReachPoints."""
+    return np.array([point.reach(k) for point in point_list])
+
+
+def _shuffle_distance(xpoint: ReachPoint, ypoint: ReachPoint, k=5) -> float:
```

Comment:
Consider reusing `Simulator._shuffle_distance` in cardinality eval fw: https://github.com/world-federation-of-advertisers/cardinality_estimation_evaluation_framework/blob/4c10fcfd6d687718850067943fdce4bd8c33d6db/src/simulations/simulator.py#L121

Reply:
I wrote that code!  As a general rule, I think it's a good practice to minimize the number of dependencies between code bases.  Since this implementation computes the shuffle distance between ReachPoints, it wouldn't be possible to eliminate this function altogether.  My feeling was that it was better to reimplement the remainder rather than introduce another dependency.  Other people in the same situation might reasonably choose a different path.

Reply:
Thanks for the explanation! I don't have enough experience on this and am very happy to learn your practice and the thought process behind it.